### PR TITLE
Use original gateway to process payment request buttons from UPE

### PR DESCRIPTION
**Pull request by bborman22**
# Changes proposed in this Pull Request:

When UPE is enabled, the payment request buttons were not working. This was due to the `process_payment` function in the UPE gateway being unable to process the payment request transactions. This issue was [addressed in the WCPay plugin](https://github.com/Automattic/woocommerce-payments/pull/2691) by falling back to the original gateway's `process_payment` function to handle any non-UPE payment methods. I have adopted the same solution here.

This PR updates the UPE gateway to fallback to the original gateway when a payment intent ID was not found during the call the process payment. This approach is essentially saying that any order UPE is unable to properly process (due to missing payment intent ID) should be handled by the original gateway.

# Testing instructions

I  have only been able to test with Google Pay, so I included those instructions here:

- Start with a site with UPE and payment request buttons enabled.
- You will need a Google Pay card set up to test this. You can check this by making sure you are logged in to pay.google.com and have a credit card on file there. To test Google Pay, you will need to be using the Chrome browser.
- Go to a product page and confirm the Google Pay button is displayed.
- Click on Google Pay and go through with the transaction.
- The order should complete successfully with this branch.
- Be sure to do thorough testing of paying with a saved payment method using UPE as well.

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

### CODE REVIEW

**Review by reykjalin — CHANGES_REQUESTED** [review]
When I do a 3DS payment in the Checkout Block (you have to use code currently in #1825) using PRBs with this change I get the following error:

```
Fatal error: Uncaught Error: Call to undefined method WC_Stripe_UPE_Payment_Gateway::verify_intent_after_checkout() in /var/www/stripe.thorlaksson.com/htdocs/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-intent-controller.php:128 Stack trace: #0 /var/www/stripe.thorlaksson.com/htdocs/wp-includes/class-wp-hook.php(303): WC_Stripe_Intent_Controller->verify_intent() #1 /var/www/stripe.thorlaksson.com/htdocs/wp-includes/class-wp-hook.php(327): WP_Hook->apply_filters() #2 /var/www/stripe.thorlaksson.com/htdocs/wp-includes/plugin.php(470): WP_Hook->do_action() #3 /var/www/stripe.thorlaksson.com/htdocs/wp-content/plugins/woocommerce/includes/class-wc-ajax.php(90): do_action() #4 /var/www/stripe.thorlaksson.com/htdocs/wp-includes/class-wp-hook.php(303): WC_AJAX::do_wc_ajax() #5 /var/www/stripe.thorlaksson.com/htdocs/wp-includes/class-wp-hook.php(327): WP_Hook->apply_filters() #6 /var/www/stripe.thorlaksson.com/htdocs/wp-includes/plugin.php(470): WP_Hook->do_action() #7 /var/www/stripe.thorlaksson.com/htdo in /var/www/stripe.thorlaksson.com/htdocs/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-intent-controller.php on line 128
There has been a critical error on this website.

Learn more about troubleshooting WordPress.
```

image

I think we can fix this by just adding a `verify_intent_after_checkout()` function to the UPE gateway class, but that might not be the best way to do this. I also don't think we can just instantiate the `WC_Gateway_Stripe` class in the `WC_Stripe_Intent_Controller` due to other payment methods.

What if the UPE gateway inherits from `WC_Gateway_Stripe`? Is that something we could do, or would that have some weird side effects that we're not so sure about? It would make it necessary to match the function signatures, but probably resolve the issue? Feels weird though.

I'm really not sure what the cleanest way to solve this problem is. Maybe by moving the `verify_intent_after_checkout()` function to the abstract class? 🤔 

***

Also; 3DS payments don't finish on the shortcode JS page because `assets/js/stripe.js` isn't loaded when UPE is enabled. This is a problem in particular because the `onHashChange()` function isn't present:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/57b9361ed8db4cca1e0dcbf9e56556e7085f885c/assets/js/stripe.js#L787-L814

My suggestion would be to add a new file like `classic/upe/legacy-support.js` and register any legacy support event handlers necessary for the PRBs to work with UPE, even though they'll be duplicated from the `assets/js/` legacy directory.

**Comment by superdav42** [comment]
> What if the UPE gateway inherits from WC_Gateway_Stripe? Is that something we could do, or would that have some weird side effects that we're not so sure about? It would make it necessary to match the function signatures, but probably resolve the issue? Feels weird though.

I think this would actually be a good idea. This is exactly what [WCPay does](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L34). Then we can use `parent::process_payment` here similar to how it's [used on WCPay](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L317) without creating a new gateway class. Instantiating a new gateway class can have side effects because the actions are all registered again. With the code is it is now you'll have two gateways which actions registered on `wp_enqueue_scripts` after `process_payment()`. Probably not an issue itself but it is an unexpected side effect when you create a gateway instance.

**Comment by bborman22** [comment]
Thanks for this feedback! I've just pushed an update to have the UPE gateway inherit from the original gateway. This more closely matches the structure from WCPay. In my testing I didn't run in to any issues with this new set up and there didn't seem to be any glaring issues that would come up from this. The one thing it required was setting default values on the `retry_after_error` function so that we didn't have to update any additional code in the UPE gateway to match the function signatures. The `process_payment` function also needed an updated signature, but defaults were already being set on the original (parent now) gateway. The other slight oddity, that I was curious for some feedback on, is that I don't call the parent constructor from the UPE gateway. I was afraid this would cause some of the unintended side effects @superdav42 mentioned, and I didn't see any need for it since the UPE constructor covers the vitals it seems, but wanted a second (or third) opinion on that.

I was also not able to test the flow of using a 3DS card within a payment request button. In my case I was testing with Google Pay and it did not allow adding the test 3DS card I had been using in other cases. So I'm unsure if this fixes the exact issue mentioned, but from following the flow it appears that it would.

> Also; 3DS payments don't finish on the shortcode JS page because assets/js/stripe.js isn't loaded when UPE is enabled. This is a problem in particular because the onHashChange() function isn't present

@reykjalin Would you be able to give more details about the payment not finishing? Was this also with a PRB in the shortcode checkout or just any 3DS in shortcode checkout? The UPE JS does have a [`hashchange` listener](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/c323a5aa6237a6ae72e5722518c157b4beca9b37/client/classic/upe/index.js#L679) that is working for the shortcode checkout when entering a 3DS card into the CC fields, but wanted to confirm where the issue might be happening.

**Comment by reykjalin** [comment]
> @reykjalin Would you be able to give more details about the payment not finishing? Was this also with a PRB in the shortcode checkout or just any 3DS in shortcode checkout? The UPE JS does have a hashchange listener that is working for the shortcode checkout when entering a 3DS card into the CC fields, but wanted to confirm where the issue might be happening.

Yes, sorry. When paying via Payment Request in the Shortcode checkout they checkout is stuck in a perpetual loading state because the `hashChange` listeners are fundamentally different; the UPE listener won't work for the PRBs.

This is not a problem when paying with a 3DS card through the UPE element, that works as expected 🙂

**Comment by reykjalin** [comment]
> The other slight oddity, that I was curious for some feedback on, is that I don't call the parent constructor from the UPE gateway. I was afraid this would cause some of the unintended side effects @superdav42 mentioned, and I didn't see any need for it since the UPE constructor covers the vitals it seems, but wanted a second (or third) opinion on that.

So long as there's no functionality missing I would prefer not to call the parent constructor. It'll help keep any side-effects to a minimum :)

**Review by reykjalin — CHANGES_REQUESTED** [review]
There are still issues with 3DS payments when using PRBs, both in the Block-based and Shortcode based checkout flows.

I was able to fix it by making relatively minor changes on my Checkout Block integration branch; you can see the changes in the attached patch -- note that some of these are already present in the current PR.

In essence, the additional changes are:

1. Add a filter to add the right redirect on a successful payment in the `WC_Stripe_UPE_Payment_Gateway` constructor
2. Add a legacy `hashchange` event handler to handle cases where the parent `process_payment()` method is used.

Note that I had to zip up the patch file for it to work as intended so you have to unzip it first:

[working-prbs.patch.zip](https://github.com/woocommerce/woocommerce-gateway-stripe/files/7207019/working-prbs.patch.zip)

You can apply the patch (after unzipping) by running

```sh
git apply working-prbs.patch
```

<details><summary>Just in case, here is the raw output from `git diff` with highlighting</summary>

```diff
diff --git a/client/classic/upe/index.js b/client/classic/upe/index.js
index 35d2fc3a..2ab7aa24 100644
--- a/client/classic/upe/index.js
+++ b/client/classic/upe/index.js
@@ -2,6 +2,7 @@ import jQuery from 'jquery';
 import WCStripeAPI from '../../api';
 import { getStripeServerData, getUPETerms } from '../../stripe-utils';
 import { getFontRulesFromPage, getAppearance } from '../../styles/upe';
+import { legacyHashchangeHandler } from './legacy-support';
 import './style.scss';
 
 jQuery( function ( $ ) {
@@ -676,9 +677,11 @@ jQuery( function ( $ ) {
 	maybeShowAuthenticationModal();
 
 	// Handle hash change - used when authenticating payment with SCA on checkout page.
-	window.addEventListener( 'hashchange', () => {
+	$( window ).on( 'hashchange', () => {
 		if ( window.location.hash.startsWith( '#wc-stripe-confirm-' ) ) {
 			maybeShowAuthenticationModal();
+		} else if ( window.location.hash.startsWith( '#confirm-' ) ) {
+			legacyHashchangeHandler( api, showError );
 		}
 	} );
 } );
diff --git a/client/classic/upe/legacy-support.js b/client/classic/upe/legacy-support.js
new file mode 100644
index 00000000..77358892
--- /dev/null
+++ b/client/classic/upe/legacy-support.js
@@ -0,0 +1,69 @@
+import $ from 'jquery';
+import { getStripeServerData } from 'wcstripe/stripe-utils';
+
+/**
+ * Handles hashchange events when using PRBs.
+ *
+ * @param {Object}   api        The Stripe API interface.
+ * @param {Function} showError  Function used to present an error to the customer.
+ */
+export const legacyHashchangeHandler = ( api, showError ) => {
+	const partials = window.location.hash.match(
+		/^#?confirm-(pi|si)-([^:]+):(.+)$/
+	);
+
+	if ( ! partials || partials.length < 4 ) {
+		return;
+	}
+
+	const type = partials[ 1 ];
+	const intentClientSecret = partials[ 2 ];
+	const redirectURL = decodeURIComponent( partials[ 3 ] );
+
+	// Cleanup the URL.
+	// https://stackoverflow.com/a/5298684
+	// eslint-disable-next-line no-undef
+	history.replaceState(
+		'',
+		document.title,
+		window.location.pathname + window.location.search
+	);
+	api.getStripe()
+		[ type === 'si' ? 'handleCardSetup' : 'handleCardPayment' ](
+			intentClientSecret
+		)
+		.then( function ( response ) {
+			if ( response.error ) {
+				throw response.error;
+			}
+
+			const intent =
+				response[ type === 'si' ? 'setupIntent' : 'paymentIntent' ];
+			if (
+				intent.status !== 'requires_capture' &&
+				intent.status !== 'succeeded'
+			) {
+				return;
+			}
+
+			window.location = redirectURL;
+		} )
+		.catch( function ( error ) {
+			$( 'form.checkout' ).removeClass( 'processing' ).unblock();
+			$( '#order_review' ).removeClass( 'processing' ).unblock();
+			$( '#payment' ).show( 500 );
+
+			let errorMessage = error.message;
+
+			// If this is a generic error, we probably don't want to display the error message to the user,
+			// so display a generic message instead.
+			if ( error instanceof Error ) {
+				errorMessage = getStripeServerData()?.genericErrorMessage;
+			}
+
+			showError( errorMessage );
+
+			// Report back to the server.
+			$.get( redirectURL + '&is_ajax' );
+		} );
+};
diff --git a/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php b/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
index 27062232..5886bb92 100644
--- a/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
+++ b/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php
@@ -7,11 +7,11 @@ if ( ! defined( 'ABSPATH' ) ) {
 /**
 * Class that handles UPE payment method.
 *
-* @extends WC_Stripe_Payment_Gateway
+* @extends WC_Gateway_Stripe
 *
 * @since 5.5.0
 */
-class WC_Stripe_UPE_Payment_Gateway extends WC_Stripe_Payment_Gateway {
+class WC_Stripe_UPE_Payment_Gateway extends WC_Gateway_Stripe {
 
 	const ID = 'stripe';
 
@@ -126,6 +126,10 @@ class WC_Stripe_UPE_Payment_Gateway extends WC_Stripe_Payment_Gateway {
 
 		add_action( 'woocommerce_update_options_payment_gateways_' . $this->id, [ $this, 'process_admin_options' ] );
 		add_action( 'wp_enqueue_scripts', [ $this, 'payment_scripts' ] );
+
+		// Needed for 3DS compatibility when checking out with PRBs..
+		// Copied from WC_Gateway_Stripe::__construct().
+		add_filter( 'woocommerce_payment_successful_result', [ $this, 'modify_successful_payment_result' ], 99999, 2 );
 	}
 
 	/**
@@ -400,7 +404,7 @@ class WC_Stripe_UPE_Payment_Gateway extends WC_Stripe_Payment_Gateway {
 	 *
 	 * @return array|null An array with result of payment and redirect URL, or nothing.
 	 */
-	public function process_payment( $order_id ) {
+	public function process_payment( $order_id, $retry = true, $force_save_source = false, $previous_error = false, $use_order_source = false ) {
 		if ( $this->is_using_saved_payment_method() ) {
 			return $this->process_payment_with_saved_payment_method( $order_id );
 		}
@@ -451,6 +455,8 @@ class WC_Stripe_UPE_Payment_Gateway extends WC_Stripe_Payment_Gateway {
 					$order
 				);
 			}
+		} else {
+			return parent::process_payment( $order_id, $retry, $force_save_source, $previous_error, $use_order_source );
 		}
 
 		return [
@@ -801,8 +807,8 @@ class WC_Stripe_UPE_Payment_Gateway extends WC_Stripe_Payment_Gateway {
 	 * @throws WC_Stripe_Exception If the payment is not accepted.
 	 * @return array|void
 	 */
-	public function retry_after_error( $intent, $order, $can_retry ) {
-		if ( ! $can_retry ) {
+	public function retry_after_error( $intent, $order, $retry, $force_save_source = false, $previous_error = false, $use_order_source = false ) {
+		if ( ! $retry ) {
 			$localized_message = __( 'Sorry, we are unable to process your payment at this time. Please retry later.', 'woocommerce-gateway-stripe' );
 			$order->add_order_note( $localized_message );
 			throw new WC_Stripe_Exception( print_r( $intent, true ), $localized_message ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.
```

</details>

**Comment by reykjalin** [comment]
I took the liberty of pushing the diff I mentioned in my previous comment (https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1908#pullrequestreview-760371022) in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1908/commits/e7c5fd8cfbc801c11fcb6d5d0437f55b192240b7 to speed up the process on this a bit 🙂

File: client/classic/upe/index.js [hunk 1]
```diff
@@ -676,9 +677,11 @@ jQuery( function ( $ ) {
 	maybeShowAuthenticationModal();
 
 	// Handle hash change - used when authenticating payment with SCA on checkout page.
-	window.addEventListener( 'hashchange', () => {
+	$( window ).on( 'hashchange', () => {
```
Comment (reykjalin) [inline client/classic/upe/index.js:680]: This is really just to get rid of a linting issue where ESLint doesn't want us to use the `window` global directly. We can change it back, this line change is definitely not necessary.

File: client/classic/upe/legacy-support.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
+import $ from 'jquery';
+import { getStripeServerData } from 'wcstripe/stripe-utils';
+
+/**
+ * Handles hashchange events when using PRBs.
+ *
+ * @param {Object}   api        The Stripe API interface.
+ * @param {Function} showError  Function used to present an error to the customer.
+ */
+export const legacyHashchangeHandler = ( api, showError ) => {
```
Comment (reykjalin) [inline client/classic/upe/legacy-support.js:10]: This is basically a copy of https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8eba0d6176c40107c12c2ebd0cfe362549bbb18d/assets/js/stripe.js#L799-L814

**Review by reykjalin — APPROVED** [review]

**Review by bborman22 — COMMENTED** [review]
Thank you so much for the additional fix and getting that merged! I was able to test now with the 3DS cards using payment requests in shortcode checkout on this branch and was able to test against #1825 with these changes with the blocks checkout and both worked. Though I had some trouble getting the shortcode checkout to work with #1825, but that might just be an oddity of how I was merging/running them?

LGTM!

**Pull request merged**